src/i2c.rs

```diff
@@ -0,0 +1,398 @@
+                        // Wait until we have received something
+                        busy_wait!(self.i2c, rxne);
+
+                        *byte = self.i2c.rxdr.read().rxdata().bits();
+                    }
+
+                    // automatic STOP
+
+                    Ok(())
+                }
+            }
+
+            impl<PINS> Read for I2c<$I2CX, PINS> {
+            type Error = Error;
+
+            fn read(
+                &mut self,
+                addr: u8,
+                buffer: &mut [u8],
+            ) -> Result<(), Error> {
+                // TODO support transfers of more than 255 bytes
+                assert!(buffer.len() < 256 && buffer.len() > 0);
+
+                // TODO do we have to explicitly wait here if the bus is busy (e.g. another
+                // master is communicating)?
+
+                // reSTART and prepare to receive bytes into `buffer`
+                self.i2c.cr2.write(|w| {
+                    w.sadd()
+                        .bits((addr << 1 | 0) as u16)
+                        .rd_wrn()
+                        .set_bit()
+                        .nbytes()
+                        .bits(buffer.len() as u8)
+                        .start()
+                        .set_bit()
+                        .autoend()
+                        .set_bit()
+                });
+
+                for byte in buffer {
+                    // Wait until we have received something
+                    busy_wait!(self.i2c, rxne);
+
+                    *byte = self.i2c.rxdr.read().rxdata().bits();
+                }
+
+                // automatic STOP
+
+                Ok(())
+            }
+            }
+        )+
+    };
+}
+
+macro_rules! pins {
+    ($($I2CX:ty: SCL: [$($SCL:ty),*] SDA: [$($SDA:ty),*])+) => {
+        $(
+            $(
+                unsafe impl PinScl<$I2CX> for $SCL {}
+            )*
+            $(
+                unsafe impl PinSda<$I2CX> for $SDA {}
+            )*
+        )+
+    }
+}
+
+pins! {
+    I2C1: 
+        SCL: [
+            PB6<Alternate<AF4>>,
+            PB8<Alternate<AF4>>
+        ]
+
+        SDA: [
+            PB7<Alternate<AF4>>,
+            PB9<Alternate<AF4>>
+        ]
+    
+    I2C2:
+        SCL: [
+            PB10<Alternate<AF4>>,
+            PF1<Alternate<AF4>>,
+            PH4<Alternate<AF4>>
+        ]
+
+        SDA: [
+            PB11<Alternate<AF4>>,
+            PF0<Alternate<AF4>>,
+            PH5<Alternate<AF4>>
+        ]
+
+    I2C3:
+        SCL: [
+            PA8<Alternate<AF4>>,
+            PH7<Alternate<AF4>>
+        ]
+
+        SDA: [
+            PC9<Alternate<AF4>>,
+            PH8<Alternate<AF4>>
+        ]
+    
+    I2C4:
+        SCL: [
+            PD12<Alternate<AF4>>,
+            PF14<Alternate<AF4>>,
+            PH11<Alternate<AF4>>,
+            PB6<Alternate<AF4>>,
+            PB8<Alternate<AF4>>
+        ]
+
+        SDA: [
+            PB7<Alternate<AF4>>,
```

Comment:
I2C4 on PB6/7/8/9 is on `AF6`

Reply:
Woops, assumed too much consistency on STMs side there, I'll double check this in the future then, thanks for the catch,